internal/batch/batch_processor.go

```diff
@@ -182,11 +191,15 @@ func (bp *batchProcessor) addWork(newWork *batchWork) (full, overflow bool) {
 	if !added {
 		newQueue = append(newQueue, newWork)
 	}
+
+	// Special handling for transactions that allow only one message per batch
+	batchOfOne := newWork.msg.Header.TxType == core.TransactionTypeContractInvokePin
```

Comment:
Note so I remember to come back to this:

I don't from this line understand how we flush the existing batch out. This makes it `full` when we find there's a `batchOfOne`... but what about when there were already some messages being assembled into a batch when this happened.

Reply:
@awrichar - I didn't find ^^^ if you could help me with where it landed.

Reply:
Yea, good catch. This was working at some point but then fell off the wagon with one of my (many) refactors of this method.

Reply:
Actually, in coming back to this - each batch processor fundamentally deals with only one type of transaction. So we don't really have to keep peeking at messages... we can just look at `bp.conf.txType` to determine if we're a _real_ batch processor or a one-item-per-batch kind of batch processor.

Reply:
OK... I think it actually would work as-is, but I pushed some logic that hopefully makes this clearer.

Essentially for a batch processor that is assigned to `TransactionTypeContractInvokePin`, the work queue will only have 1 item (full but not overflowing) or 2 items (full and overflowed). From there I think the existing logic will take care of flushing it properly. Correct me if I'm missing anything though.

Reply:
Great! I spent some time re-familiarizing myself with the logic for `overflow` (where later we retain the overflowed item for the next batch) - that was definitely what I was missing when I made the comment 👍